coreMacros/src/main/scala/Chisel/internal/sourceinfo/SourceInfoTransform.scala

```diff
@@ -0,0 +1,180 @@
+  }
+
+  def elsewhen(elseCond: c.Tree)(block: c.Tree): c.Tree = {
+    q"$thisObj.do_elsewhen($elseCond, $block)($implicitSourceInfo)"
+  }
+
+  def otherwise(block: c.Tree): c.Tree = {
+    q"$thisObj.do_otherwise($block)($implicitSourceInfo)"
+  }
+}
+
+class StopTransform(val c: Context) extends SourceInfoTransformMacro {
+  import c.universe._
+  def stop(): c.Tree = {
+    q"$thisObj.do_stop($implicitSourceInfo)"
+  }
+}
+
+class PrintTransform(val c: Context) extends SourceInfoTransformMacro {
+  import c.universe._
+  def apply(fmt: c.Tree, data: c.Tree*): c.Tree = {
+    q"$thisObj.do_apply($fmt, ..$data)($implicitSourceInfo)"
+  }
+}
+
+class AssertTransform(val c: Context) extends SourceInfoTransformMacro {
+  import c.universe._
+  def apply_msg(cond: c.Tree, message: c.Tree): c.Tree = {
+    q"$thisObj.do_apply($cond, $message)($implicitSourceInfo)"
+  }
+  def apply(cond: c.Tree): c.Tree = {
+    q"$thisObj.do_apply($cond)($implicitSourceInfo)"
+  }
+}
+
+class MuxTransform(val c: Context) extends SourceInfoTransformMacro {
+  import c.universe._
+  def apply[T: c.WeakTypeTag](cond: c.Tree, con: c.Tree, alt: c.Tree): c.Tree = {
+    val tpe = weakTypeOf[T]
+    q"$thisObj.do_apply[$tpe]($cond, $con, $alt)($implicitSourceInfo)"
+  }
+}
+
+class VecTransform(val c: Context) extends SourceInfoTransformMacro {
+  import c.universe._
+  def apply_elts(elts: c.Tree): c.Tree = {
+    q"$thisObj.do_apply($elts)($implicitSourceInfo)"
+  }
+  def apply_elt0(elt0: c.Tree, elts: c.Tree*): c.Tree = {
+    q"$thisObj.do_apply($elt0, ..$elts)($implicitSourceInfo)"
+  }
+  def tabulate(n: c.Tree)(gen: c.Tree): c.Tree = {
+    q"$thisObj.do_tabulate($n)($gen)($implicitSourceInfo)"
+  }
+  def fill(n: c.Tree)(gen: c.Tree): c.Tree = {
+    q"$thisObj.do_fill($n)($gen)($implicitSourceInfo)"
+  }
+  def contains(x: c.Tree)(ev: c.Tree): c.Tree = {
+    q"$thisObj.do_contains($x)($implicitSourceInfo, $ev)"
+  }
+}
+
+/** "Automatic" source information transform / insertion macros, which generate the function name
+  * based on the macro invocation (instead of explicitly writing out every transform).
+  */
+class SourceInfoTransform(val c: Context) extends SourceInfoTransformMacro {
+  import c.universe._
+  /** Returns the TermName of the transformed function, which is the applied function name with do_
+    * prepended.
+    */
+  def doFuncTerm = {
+    val funcName = c.macroApplication match {
+      case Apply(TypeApply(Select(_, TermName(funcName)), _), _) => funcName
+      case Apply(Select(_, TermName(funcName)), _) => funcName
+      case Select(_, TermName(funcName)) => funcName
+
+    }
+    TermName("do_" + funcName)
+  }
+
+  /** Transforms a call of the form obj.fn(that) into obj.do_fn(sourceInfo, that). Single argument
+    * must be named that.
+    */
+  def noArg(): c.Tree = {
+    q"$thisObj.$doFuncTerm($implicitSourceInfo)"
+  }
+
+  def thatArg(that: c.Tree): c.Tree = {
+    q"$thisObj.$doFuncTerm($that)($implicitSourceInfo)"
+  }
+
```

Comment:
Should probably be a separate pull request, preferably after this one to avoid merge conflicts. Open an issue with suggested resolution?